backend/src/main/java/com/example/backend/Specifications/ReservationSpecifications.java

```diff
@@ -0,0 +1,129 @@
+        return (root, query, criteriaBuilder) -> {
+            return criteriaBuilder.equal(root.get("userId"), userId);
+        };
+    }
+
+    public static Specification<Reservation> containsSource(String source) {
+        return (root, query, criteriaBuilder) -> {
+
+            Join<Reservation, FlightLeg> flightLegJoin = root.join("flight").join("flightLegs");
+            Join<FlightLeg, Airport> airportJoin = flightLegJoin.join("departureAirport");
+
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightLegJoin.get("flightLegId"), 1),
+                    criteriaBuilder.equal(airportJoin.get("airportCity"), source)
+            );
+        };
+    }
+
+    public static Specification<Reservation> containsDestination(String destination) {
+        return (root, query, criteriaBuilder) -> {
+
+            Join<Reservation, Flight> flightJoin = root.join("flight");
+
+            Subquery<Integer> subquery = query.subquery(Integer.class);
+            Root<FlightLeg> subqueryRoot = subquery.from(FlightLeg.class);
+
+            subquery.select(criteriaBuilder.max(subqueryRoot.get("flightLegId")))
+                    .where(criteriaBuilder.equal(subqueryRoot.get("flightId"), flightJoin.get("id")))
+                    .groupBy(subqueryRoot.get("flightId"));
+
+            Join<Flight, FlightLeg> flightLegJoin = flightJoin.join("flightLegs");
+            Join<FlightLeg, Airport> airportJoin = flightLegJoin.join("arrivalAirport");
+
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightLegJoin.get("flightLegId"), subquery),
+                    criteriaBuilder.equal(airportJoin.get("airportCity"), destination)
+            );
+        };
+    }
+
+    public static Specification<Reservation> containsDepartureDate(LocalDate departureDate) {
+        return (root, query, criteriaBuilder) -> {
+
+            var flightJoin = root.join("flight");
+
+            return criteriaBuilder.equal(flightJoin.get("departureDate"), departureDate);
+        };
+    }
+
+    public static Specification<Reservation> sortedByDepartureDate(String direction) {
+        return (root, query, criteriaBuilder) -> {
+
+            var flightJoin = root.join("flight");
+
+            if (direction != null && !direction.isEmpty()) {
+                Sort.Direction sortDirection = Sort.Direction.fromString(direction);
+                query.orderBy(sortDirection == Sort.Direction.ASC
+                        ? criteriaBuilder.asc(flightJoin.get("departureDate"))
+                        : criteriaBuilder.desc(flightJoin.get("departureDate")));
+            }
+
+            return criteriaBuilder.conjunction();
+        };
+    }
+
+    public static Specification<Reservation> sortedByArrivalDate(String direction) {
+        return (root, query, criteriaBuilder) -> {
+
+            var flightJoin = root.join("flight");
+
```

Comment:
In this file there are many duplicated blocks of code for example here in the two methods sortedByDepartureDate  and sortedByArrivalDate they are identical except the attribute you passing to the get function.
So please break functions into small reusable functions.

Reply:
There was many duplicated code as you mentioned, so I combined them into functions without much duplicates.